### ISSUE TITLE
[ENH] refactoring html representation documentation link to allow custom linking in private and public extensions to sktime

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#6911]
in https://github.com/sktime/sktime/pull/6876, @mateuszkasprowicz has added functionality (similar to `sklearn`) in the html repr of estimator objects to link to the documentation.

This covers `sktime` native estimators, but the link will point to a non-existent location for 2nd/3rd party packages (public packages using the `sktime` interface), as well as private packages.

I think we should give power users and extenders an option to specify the location the documentation link points to, I would approach the design in two levels here:

* a tag, e.g., `doc_loc`, that is a str with an URL. This would be placed in the "simple" extension templates.
* the option to override `_get_doc_link`, as an intentional part of the extension template. This should only be used if there are many estimators in the same package, e.g., following a documentation pattern. The advanced (and not the simple) extension templates, and/or the documentation of the `doc_loc` tag could explain this.

Finally, once the `get_doc_link` override point is established, we want to do the following refactor:

* move the mixin up to `scikit-base`
* change the `sktime` mixin to be just an override of the `get_doc_link` method or similar.

FYI @mateuszkasprowicz as the author; FYI @felipeangelimvieira, @tveten as the two most likely individuals known to me who may like to link to separate documentation online.

**Comment by fkiraly** [comment]
Any thoughts about the design, please comment.

**Comment by mateuszkasprowicz** [comment]
Maybe first a (hopefully) helpful note about the current implementation. To create a valid documentation link you need following arguments:

1. documentation link template - a structure of a documentation link 
`f"https://www.sktime.net/en/v{version}/api_reference/auto_generated/{object_path}.html"`
2. root module name - this is to get a version of the library a user has, i.e. {version}, e.g., "sktime" or "sklearn"
3. object path in docs - a function that takes object's self, and outputs a valid object docs path, i.e. {object_path} (`_doc_link_url_param_generator` in the current implementation)

For @darksapien23151:
Currently, the logic for getting object path is split between `_get_reduced_path` and `get_doc_link`, so we may want to create a separate function and pass it to `_doc_link_url_param_generator` e.g., 
```python
def get_doc_path(obj):
    modpath = str(self.__class__)[8:-2]
    doc_path = _get_reduced_path(modpath)
    return doc_path
```
  In retrospect, I'd say we can also think of moving the standalone functions `_write_base_object_html`, `_write_label_html`, and `_get_reduced_path` to `_HTMLDocumentationLinkMixin`. 


### About the new design

Tags sound like a promising solution. Doc link template and root module name are good candidates to be tags. But I see some challenges with handling `_doc_link_url_param_generator` which should be a function. 

What comes to my mind is to specify an import path to  `_doc_link_url_param_generator` and then import and call inside `_HTMLDocumentationLinkMixin`. Feels hacky but should work.

`get_doc_link` method I wouldn't expose for user to change. Currently, it serves mostly as a dispatcher i.e., 

1. if class module is not from the specified `_doc_link_module` return "", so we don't show the docs icon for them
https://github.com/sktime/sktime/blob/5e6d8a46c4c231d3c3a474095f5907c16e10da49/sktime/utils/_estimator_html_repr.py#L84-L91
2. else if we have not specified `_doc_link_url_param_generator`, there's some reasonable default
3. else we return documentation link template formatted with `_doc_link_url_param_generator`

**Comment by fkiraly** [comment]
@mateuszkasprowicz, thanks for the update!

FYI @felipeangelimvieira re prophetverse.

For @darksapien23151, what we usually expect for an issue like this is a proposed design, which is now being discussed after @mateuszkasprowicz´s post.
I did not notice you either, possibly lost in my inbox, you could ping on the discord dev-chat in such a situation.

I will reply about the design after reading in more detail.

**Comment by fkiraly** [comment]
@mateuszkasprowicz, very interesting analysis.

Let me respond organized by the three pieces of data.

data 2 - library and version.

We have noticed in other places that this is relevant - not just for docs, but also for persistence (save/load), where encoding a pointer to the library and its version also seems crucial in various designs.

Given that this appears in multiple places, I think the long-term solution should live in `scikit-base`, with some programmatic way to get root library and version. Optionally, perhaps that of any dependencies of the estimator as well, or even a full pip freeze snapshot.

Mid- and short-term, the modular surface does not yet exist, but I would prefer introducing more general tags, for instance `source_pkg` and `source_pkg_version`, how about this?

data 1 plus 3 - question: this seems a bit restrictive, should we not allow general paths to be generated per object? That would result in a single datum instead of two.

I also do not like setting functions as attributes as in options 1 and 3, this is a python anti-pattern.

How about having a default implementation of *a method* ` _get_link_url_params` in `scikit-base`, and allowing users to override it? Using a more restrictive pattern as above can be defined at package level.

**Comment by mateuszkasprowicz** [comment]
ad. 2
About the change "doc_link_module" -> "source_pkg" - agreed. But the version as a tag - not sure. Currently, the solution relies on the `__version__` defintion in the root module which to me is reasonable. https://github.com/sktime/sktime/blob/1f4bd06c8261418969b505a9f7da46339095b84b/sktime/utils/_estimator_html_repr.py#L286-L288
Then, do we want the tag value to be generated dynamically?

ad. 1&3
Well, even if you dumped a template altogether, you would still need some link structure defined inside the function e.g.,
```python
# simplifed
def _get_link_url_params(base_object):
    object_path = str(base_object.__class__)[8:-2]
    return f"https://www.sktime.net/en/api_reference/auto_generated/{object_path}.html"
```
I'm not loving it, but I can imagine it gives more flexibility when someone has a messy docs structure (then you can do some clever if/elif based on object path).

About having this as a method - makes sense.

**Comment by fkiraly** [comment]
> ad. 2
> Then, do we want the tag value to be generated dynamically?

Currently it is generated dynamically in `sktime`, to ensure that exported models (serialized etc) have the tag with the version under which it was exported. See line 88 in `sktime.base._base`.

> I'm not loving it,

It seems there are only imperfect solutions here. Personally, I am weighing the "lower number of API points" higher than customizability, since that seems easier for someone who has not designed the interface to understand and comply with.

How exactly we reduce the number of interface points is not that important for me though.

> With all three solutions, every time they would need to override `_get_link_url_params` which seems redundant

Hm, why is that better in which alternative, could you kindly explain? Extensibility is a very strong design desideratum for me, but I am not grokking the point here (e.g., which alternatives, which better and why).

**Comment by mateuszkasprowicz** [comment]
I see your point here - so to summarize the approach:

1. `source_pkg` and `source_pkg_version` as tags -> for doc link we only really need `source_pkg_version`
2. implement `_get_link_url_params` method (or maybe `_generate_doc_link`) which is takes self as an input and returns a valid doc link to every object e.g.,
```python
# simplifed
def _generate_doc_link(self):
    object_path = str(self.__class__)[8:-2]
    version = self.get_tag("source_pkg_version")
    return f"https://www.sktime.net/en/{version}/api_reference/auto_generated/{object_path}.html"
```
this should be overridden in 2nd/3rd party packages.

is this correct?

**Comment by fkiraly** [comment]
@mateuszkasprowicz, yes - with the added point that `_generate_doc_link` could itself add another template pattern internally, that, in a larger package like `sktime`, could be overridden again.

I think this makes the API very simple.
